src/EventSourcing/Domain/DomainEvent.php

```diff
@@ -0,0 +1,16 @@
+<?php
+
+declare(strict_types=1);
+
+namespace EventSourcingWorkshop\EventSourcing\Domain;
+
+use DateTimeImmutable;
+
+/** @psalm-immutable */
+interface DomainEvent
+{
+    public function raisedAt(): DateTimeImmutable;
+
+    /** @return array<non-empty-string, string|int|bool|float|array|null> */
+    public function toArray(): array;
```

Comment:
I think `JsonSerializable` may be a better API definition here (to be inherited from), although I really don't like the idea of a `DomainEvent` extending an SPL object :thinking:

Reply:
`toArray()` may still be the best solution, with documentation explaining that the parameters **must** match the constructor arguments

Reply:
For now, sticking with `toArray()`, although the serialization will need some work.